File: apps/api/src/agentforge_api/models/execution.py
# apps/api/src/agentforge_api/models/execution.py

"""Execution domain models."""

from datetime import datetime
from enum import Enum
from typing import Annotated, Any, NewType

from pydantic import BaseModel, Field


ExecutionId = NewType("ExecutionId", str)


class ExecutionStatus(str, Enum):
    """Overall execution status for the entire workflow run."""
    
    PENDING = "pending"       # Created, not yet started
    RUNNING = "running"       # At least one node is executing
    COMPLETED = "completed"   # All nodes finished successfully
    FAILED = "failed"         # At least one node failed (after retries)
    CANCELLED = "cancelled"   # User-initiated cancellation


class NodeExecutionStatus(str, Enum):
    """Execution status for a single node."""
    
    PENDING = "pending"     # Waiting for dependencies
    QUEUED = "queued"       # Dependencies met, in queue
    RUNNING = "running"     # Currently executing
    COMPLETED = "completed" # Finished successfully
    FAILED = "failed"       # Failed after all retries
    SKIPPED = "skipped"     # Skipped due to upstream failure


class NodeExecutionState(BaseModel, frozen=True):
    """Runtime state of a single node during execution."""
    
    node_id: str
    status: NodeExecutionStatus = NodeExecutionStatus.PENDING
    
    started_at: datetime | None = None
    completed_at: datetime | None = None
    
    retry_count: Annotated[int, Field(ge=0)] = 0
    error: str | None = None
    output: Any | None = None


class Execution(BaseModel, frozen=True):
    """A single execution run of a workflow."""
    
    id: Annotated[str, Field(description="Unique execution identifier")]
    workflow_id: str
    status: ExecutionStatus = ExecutionStatus.PENDING
    
    # Snapshot of workflow version at execution time
    workflow_version: Annotated[int, Field(ge=1)]
    
    # User who triggered the execution
    triggered_by: str
    
    # Execution timing
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    
    # Per-node execution state
    node_states: list[NodeExecutionState] = Field(default_factory=list)
    
    # Inputs provided at execution start
    inputs: dict[str, Any] = Field(default_factory=dict)
    
    @property
    def execution_id(self) -> ExecutionId:
        """Return typed ExecutionId."""
        return ExecutionId(self.id)
    
    def get_node_state_map(self) -> dict[str, NodeExecutionState]:
        """Build node state lookup map."""
        return {state.node_id: state for state in self.node_states}

File: packages/contracts/src/types/execution.ts
// src/types/execution.ts

import type { WorkflowId } from "./workflow";
import type { NodeId } from "./node";

/**
 * Unique identifier for an execution run.
 */
export type ExecutionId = string & { readonly __brand: "ExecutionId" };

/**
 * Overall execution status for the entire workflow run.
 */
export type ExecutionStatus =
  | "pending" // Created, not yet started
  | "running" // At least one node is executing
  | "completed" // All nodes finished successfully
  | "failed" // At least one node failed (after retries)
  | "cancelled"; // User-initiated cancellation

/**
 * Execution status for a single node.
 */
export type NodeExecutionStatus =
  | "pending" // Waiting for dependencies
  | "queued" // Dependencies met, in queue
  | "running" // Currently executing
  | "completed" // Finished successfully
  | "failed" // Failed after all retries
  | "skipped"; // Skipped due to upstream failure

/**
 * Runtime state of a single node during execution.
 */
export interface NodeExecutionState {
  readonly nodeId: NodeId;
  readonly status: NodeExecutionStatus;

  /** When this node started executing (null if not started) */
  readonly startedAt: string | null;

  /** When this node finished (null if not finished) */
  readonly completedAt: string | null;

  /** Number of retry attempts made (0 = first attempt) */
  readonly retryCount: number;

  /** Error message if failed (null otherwise) */
  readonly error: string | null;

  /** Output data if completed (null otherwise) */
  readonly output: unknown | null;
}

/**
 * A single execution run of a workflow.
 */
export interface Execution {
  readonly id: ExecutionId;
  readonly workflowId: WorkflowId;
  readonly status: ExecutionStatus;

  /** Snapshot of workflow version at execution time */
  readonly workflowVersion: number;

  /** User who triggered the execution */
  readonly triggeredBy: string;

  /** Execution timing */
  readonly createdAt: string;
  readonly startedAt: string | null;
  readonly completedAt: string | null;

  /** Per-node execution state, keyed by NodeId */
  readonly nodeStates: ReadonlyMap<NodeId, NodeExecutionState>;

  /** Inputs provided at execution start */
  readonly inputs: Record<string, unknown>;
}
